src/libraries/Common/src/System/Net/IPv6AddressHelper.Common.cs

```diff
@@ -272,99 +298,105 @@ internal static unsafe bool IsValidStrict(char* name, int start, ref int end)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            int number = 0;
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
+            ushort number = 0;
+            int currentCh;
             int index = 0;
             int compressorIndex = -1;
             bool numberIsValid = true;
 
-            //This used to be a class instance member but have not been used so far
-            int PrefixLength = 0;
-            if (address[start] == '[')
+            scopeId = ReadOnlySpan<TChar>.Empty;
+            // Skip the start '[' character, if present. Stop parsing at the end IPv6 address terminator (']').
+            for (int i = (address[0] == TChar.CreateTruncating('[') ? 1 : 0); i < address.Length && address[i] != TChar.CreateTruncating(']');)
             {
-                ++start;
-            }
+                currentCh = int.CreateTruncating(address[i]);
 
-            for (int i = start; i < address.Length && address[i] != ']';)
-            {
-                switch (address[i])
+                switch (currentCh)
                 {
                     case '%':
                         if (numberIsValid)
                         {
-                            numbers[index++] = (ushort)number;
+                            numbers[index++] = number;
                             numberIsValid = false;
                         }
 
-                        start = i;
-                        for (++i; i < address.Length && address[i] != ']' && address[i] != '/'; ++i)
+                        // The scope follows a '%' and terminates at the natural end of the address, the address terminator (']') or the prefix delimiter ('/').
+                        int scopeStart = i;
+
+                        for (++i; i < address.Length && address[i] != TChar.CreateTruncating(']') && address[i] != TChar.CreateTruncating('/'); ++i)
                         {
                         }
-                        scopeId = new string(address.Slice(start, i - start));
-                        // ignore prefix if any
-                        for (; i < address.Length && address[i] != ']'; ++i)
+                        scopeId = address.Slice(scopeStart, i - scopeStart);
+
+                        // Ignore the prefix (if any.)
+                        for (; i < address.Length && address[i] != TChar.CreateTruncating(']'); ++i)
                         {
                         }
                         break;
 
                     case ':':
-                        numbers[index++] = (ushort)number;
+                        numbers[index++] = number;
                         number = 0;
+                        // Two sequential colons form a compressor ('::').
                         ++i;
-                        if (address[i] == ':')
+                        if (i < address.Length && address[i] == TChar.CreateTruncating(':'))
```

Comment:
Why did this get an i < address.Length check where there wasn't one before?

Reply:
It was a remnant of the previous effort to eliminate some bounds checks on the ReadOnlySpan. It's no longer needed, so I've removed it.